daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -151,27 +151,41 @@ object TransactionCoder {
         }
 
       case e: NodeExercises[Nid, Cid, Val] =>
-        encodeVal(e.chosenValue).map {
-          case (vversion, arg) =>
-            val exBuilder =
-              TransactionOuterClass.NodeExercise
-                .newBuilder()
-                .setChoice(e.choiceId)
-                .setTemplateId(ValueCoder.encodeIdentifier(e.templateId, Some(vversion))._2)
-                .setChosenValue(arg)
-                .setConsuming(e.consuming)
-                .setContractIdOrStruct(encodeCid, transactionVersion, e.targetCoid)(
-                  _.setContractId(_),
-                  _.setContractIdStruct(_))
-                .addAllActors(e.actingParties.map(_.underlyingString).asJava)
-                .addAllChildren(e.children.map(encodeNid).toList.asJava)
-                .addAllControllers(e.controllers.map(_.underlyingString).asJava)
-                .addAllSignatories(e.signatories.map(_.underlyingString).asJava)
-                .addAllStakeholders(e.stakeholders.map(_.underlyingString).asJava)
+        for {
+          argValue <- encodeVal(e.chosenValue)
+          retValue <- e.exerciseResult
+            .map(er => encodeVal(er).map(Some(_)))
+            .getOrElse(
+              Right[EncodeError, Option[(ValueVersion, ValueOuterClass.VersionedValue)]](None))
+        } yield {
+          val (vversion, arg) = argValue
+          val exBuilder =
+            TransactionOuterClass.NodeExercise
+              .newBuilder()
+              .setChoice(e.choiceId)
+              .setTemplateId(
+                ValueCoder.encodeIdentifier(e.templateId, Some(vversion))._2
+              )
+              .setChosenValue(arg)
+              .setConsuming(e.consuming)
+              .setContractIdOrStruct(encodeCid, transactionVersion, e.targetCoid)(
+                _.setContractId(_),
+                _.setContractIdStruct(_)
+              )
+              .addAllActors(e.actingParties.map(_.underlyingString).asJava)
+              .addAllChildren(e.children.map(encodeNid).toList.asJava)
+              .addAllControllers(e.controllers.map(_.underlyingString).asJava)
+              .addAllSignatories(e.signatories.map(_.underlyingString).asJava)
+              .addAllStakeholders(e.stakeholders.map(_.underlyingString).asJava)
+          if (transactionVersion precedes minExerciseResult) {
+            if (e.exerciseResult.nonEmpty)
+              Left(EncodeError(transactionVersion, isTooOldFor = "NodeExercise result value"))
+            else
+              Right(nodeBuilder.setExercise(exBuilder.setReturnValue(retValue.get._2)))
+          }
```

Comment:
We're going with @bitonic's plan elsewhere so when precedes, ignore `retValue`, otherwise fold it (do not use `get` or `nonEmpty`). Also this expression has to occur as the result of the whole containing block; you're just dropping the `Left` or `Right` on the floor right now.